### ISSUE TITLE
OutputDebugString prints twice

### ISSUE THREAD

**Issue by Nukem9** [x64dbg/x64dbg#224]
Output in the log looks like this:

```
#!

DebugString: "Foliage: --------------------------\n"
DebugString: "Foliage: --------------------------\n"
DebugString: "Foliage: smodelCount = 0\n"
DebugString: "Foliage: smodelCount = 0\n"
DebugString: "Foliage: dyn smodel count = 0 (max=2048)\n"
DebugString: "Foliage: dyn smodel count = 0 (max=2048)\n"
DebugString: "Foliage: sizeof(g_activeModelsHash) = 1152, sizeof(g_fullySquashedModelBits) = 256\n"
DebugString: "Foliage: sizeof(g_activeModelsHash) = 1152, sizeof(g_fullySquashedModelBits) = 256\n"
DebugString: "Foliage: --------------------------\n\n"
DebugString: "Foliage: --------------------------\n\n"
DebugString: "Error: Could not load xmodel \"skybox_mp_hotel\".\n"
DebugString: "Error: Could not load xmodel \"skybox_mp_hotel\".\n"
DebugString: "vertex stream 2: --------------------------\n"
DebugString: "vertex stream 2: --------------------------\n"
DebugString: "vertex stream 2: needs 0 bytes\n"
DebugString: "vertex stream 2: needs 0 bytes\n"
DebugString: "vertex stream 2: already allocated 0 bytes\n"
DebugString: "vertex stream 2: already allocated 0 bytes\n"
DebugString: "vertex stream 2: --------------------------\n"
DebugString: "vertex stream 2: --------------------------\n"
DebugString: "Error: Could not load rawfile \"exposure/mp_ui_viewer.d3dbsp.xpo\".\n"
DebugString: "Error: Could not load rawfile \"exposure/mp_ui_viewer.d3dbsp.xpo\".\n"
DebugString: "Error: couldn't open 'exposure/mp_ui_viewer.d3dbsp.xpo'.\n"
DebugString: "Error: couldn't open 'exposure/mp_ui_viewer.d3dbsp.xpo'.\n"
DebugString: "Error: Could not load rawfile \"exposure/default.xpo\".\n"
DebugString: "Error: Could not load rawfile \"exposure/default.xpo\".\n"
DebugString: "Error: couldn't open 'exposure/default.xpo'. This is a default file that you should have.\n"
DebugString: "Error: couldn't open 'exposure/default.xpo'. This is a default file that you should have.\n"
DebugString: "Com_TouchMemory: 0 msec. Using sum: 0\n"
DebugString: "Com_TouchMemory: 0 msec. Using sum: 0\n"
```

----------------------------------------
- Bitbucket: https://bitbucket.org/mrexodia/x64_dbg/issue/224
- Originally reported by: [Nukem9](http://bitbucket.org/Nukem9)
- Originally created at: 2014-12-21T03:11:21.683

**Comment by Nukem9** [comment]
Do you have an executable to reproduce this? And on which OS did you try?

I used the following code:

```
#!c++
#include <windows.h>

int main()
{
    OutputDebugStringA("This is a test...");
    OutputDebugStringA("Another debug string\n");
    return 0;
}
```

And I got the following result:
```
DebugString: "This is a test..."
DebugString: "Another debug string\n"
```

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
I can reproduce it reliably with x64dbg, but it works normally on Ollydbg v2. No exe because it will only run on Steam if you own the game.

I'll try to figure this out myself for now.

----------------------------------------
Original comment by: [Nukem9](http://bitbucket.org/Nukem9)

**Comment by Nukem9** [comment]
The reason could be that exceptions are not marked as handled, they are
passed to the debuggee.

This is to prevent people throwing DBGPRINT exceptions and then detecting
the debugger.

A way would be to only show first chance exception debug strings, but I
think this information is not available (api is not transparent to the fact
debug events and RIP events are actually exceptions).

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
edit: you beat me anyway
Found out why: http://maximumcrack.wordpress.com/2009/06/22/outputdebugstring-awesomeness/

```
OUTPUT_DEBUG_STRING_EVENT behaves just like an exception event, so that if you continue with DBG_EXCEPTION_NOT_HANDLED, it sends the event a second time, just like a second chance exception.

Recently I’ve been wondering why I got 2 debug string events for each call to OutputDebugString from the debuggee. I was looking for all sorts of weird solutions until I found out that returning DBG_CONTINUE did the trick.
```

https://bitbucket.org/mrexodia/titanengine-update/src/29f8973700c61a53f36f8f8247ea3ae3f909202a/TitanEngine/TitanEngine.Debugger.DebugLoop.cpp?at=master#cl-466

----------------------------------------
Original comment by: [Nukem9](http://bitbucket.org/Nukem9)

**Comment by Nukem9** [comment]
indeed, but this was done with a reason. Look for OutputDebugString exception anti debug.

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
See this issue for more information on why I throw the exception to the debuggee: issue #211: int3 (debug interrupt 'CC') not easily skipped

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
DBG: resolved issue #224 (print debug strings twice) + StringUtils::Escape function instead of separate code everywhere
→ <<cset 3378918d1311>>

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)
